### PR TITLE
Add CLIPSeg

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR adds CLIPSeg, a nice extension of CLIP for zero-shot and one-shot (image-guided) image segmentation.

To do:

- [x] transfer checkpoints and update code
- [x] update base_model_prefix

### CODE REVIEW

File: src/transformers/models/clipseg/test.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
+import torch
```
Comment (NielsRogge) [inline src/transformers/models/clipseg/test.py:1]: Yes I'll remove this script

File: docs/source/en/model_doc/clipseg.mdx [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+specific language governing permissions and limitations under the License.
+-->
+
+# CLIPSeg
+
+## Overview
+
+The CLIPSeg model was proposed in [Image Segmentation Using Text and Image Prompts](<INSERT PAPER LINK HERE>) by Timo Lüddecke
```
Comment (sgugger) [inline docs/source/en/model_doc/clipseg.mdx:17]: Paper link missing.

File: src/transformers/models/clipseg/configuration_clipseg.py [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+    >>> config = CLIPSegConfig.from_text_vision_configs(config_text, config_vision)
+    ```"""
+
+    model_type = "clipseg"
+    is_composition = True
+
+    def __init__(
+        self,
+        text_config_dict=None,
+        vision_config_dict=None,
```
Comment (sgugger) [inline src/transformers/models/clipseg/configuration_clipseg.py:321]: Make sure you use the latest improvements made in the CLIPConfig here.

File: src/transformers/models/clipseg/configuration_clipseg.py [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+            `Dict[str, any]`: Dictionary of all the attributes that make up this configuration instance,
+        """
+        output = copy.deepcopy(self.__dict__)
+        output["text_config"] = self.text_config.to_dict()
+        output["vision_config"] = self.vision_config.to_dict()
+        output["model_type"] = self.__class__.model_type
+        return output
+
+
+class CLIPSegOnnxConfig(OnnxConfig):
```
Comment (sgugger) [inline src/transformers/models/clipseg/configuration_clipseg.py:385]: I'd leave the ONNX export for a followup PR.

File: src/transformers/models/clipseg/modeling_clipseg.py [hunk 5]
```diff
@@ -0,0 +1,1503 @@
...
+        if not return_dict:
+            return tuple(
+                v
+                for v in [
+                    output,
+                    all_hidden_states,
+                    all_attentions,
+                ]
+                if v is not None
+            )
```
Comment (sgugger) [inline src/transformers/models/clipseg/modeling_clipseg.py:1323]: This should fit in one line.

File: src/transformers/models/clipseg/test.py (hunk 1 above)
Comment (sgugger) [reply src/transformers/models/clipseg/test.py:1]: If you know it shouldn't be in the PR, why add it? 🤷‍♂️

File: tests/models/clipseg/test_modeling_clipseg.py [hunk 6]
```diff
@@ -0,0 +1,743 @@
...
+
+@require_torch
+class CLIPSegModelTest(ModelTesterMixin, unittest.TestCase):
+    all_model_classes = (
+        (
+            CLIPSegModel,
+            CLIPSegForImageSegmentation,
+        )
+        if is_torch_available()
+        else ()
```
Comment (sgugger) [inline tests/models/clipseg/test_modeling_clipseg.py:417]: Should all fit in one line.

**Review by sgugger — COMMENTED** [review]
Looking good, thanks for adding this model! Make sure you update the paper link in the doc file, and the config should take into account the latest modifications done on the CLIP config.

File: tests/models/clipseg/test_processor_clipseg.py [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+
+    def prepare_image_inputs(self):
+        """This function prepares a list of PIL images, or a list of numpy arrays if one specifies numpify=True,
+        or a list of PyTorch tensors if one specifies torchify=True.
+        """
+
+        image_inputs = [np.random.randint(255, size=(3, 30, 400), dtype=np.uint8)]
+
+        image_inputs = [Image.fromarray(np.moveaxis(x, 0, -1)) for x in image_inputs]
+
```
Comment (alaradirik) [inline tests/models/clipseg/test_processor_clipseg.py:88]:
```suggestion
        """
        This function prepares a list of PIL images, or a list of numpy arrays if one specifies numpify=True,
        or a list of PyTorch tensors if one specifies torchify=True.
        """
        image_inputs = [np.random.randint(255, size=(3, 30, 400), dtype=np.uint8)]
        image_inputs = [Image.fromarray(np.moveaxis(x, 0, -1)) for x in image_inputs]

```

File: src/transformers/models/clipseg/configuration_clipseg.py [hunk 8]
```diff
@@ -0,0 +1,429 @@
...
+    r"""
+    This is the configuration class to store the configuration of a [`CLIPSegModel`]. It is used to instantiate an
+    CLIPSeg model according to the specified arguments, defining the model architecture. Instantiating a configuration
+    with the defaults will yield a similar configuration to that of the CLIPSeg
+    [organization/clipseg-rd64-uni](https://huggingface.co/organization/clipseg-rd64-uni) architecture.
+
+    Configuration objects inherit from [`PretrainedConfig`] and can be used to control the model outputs. Read the
+    documentation from [`PretrainedConfig`] for more information.
+
+
```
Comment (alaradirik) [inline src/transformers/models/clipseg/configuration_clipseg.py:49]: Extra blank line over here
```suggestion
```

File: src/transformers/models/clipseg/configuration_clipseg.py [hunk 9]
```diff
@@ -0,0 +1,429 @@
...
+        decoder_attention_dropout (`float`, *optional*, defaults to 0.0):
+            The dropout ratio for the attention probabilities.
+        decoder_hidden_act (`str` or `function`, *optional*, defaults to `"quick_gelu"`):
+            The non-linear activation function (function or string) in the encoder and pooler. If string, `"gelu"`,
+            `"relu"`, `"selu"` and `"gelu_new"` ``"quick_gelu"` are supported. layer_norm_eps (`float`, *optional*,
+            defaults to 1e-5): The epsilon used by the layer normalization layers.
+        decoder_intermediate_size (`int`, *optional*, defaults to 2048):
+            Dimensionality of the "intermediate" (i.e., feed-forward) layers in the Transformer decoder.
+        conditional_layer (`int`, *optional*, defaults to 0):
+            ...
```
Comment (alaradirik) [inline src/transformers/models/clipseg/configuration_clipseg.py:285]: This line is left incomplete, it's just the number of conditional layers, right?

File: src/transformers/models/clipseg/modeling_clipseg.py [hunk 10]
```diff
@@ -0,0 +1,1503 @@
...
+            for k in self.keys()
+        )
+
+
+@dataclass
+class CLIPSegDecoderOutput(ModelOutput):
+    """
+    Args:
+        predicted_masks (`torch.FloatTensor` of shape `(1,)`, *optional*, returned when `return_loss` is `True`):
+            ...
```
Comment (alaradirik) [inline src/transformers/models/clipseg/modeling_clipseg.py:123]: Could you add the description?

**Review by alaradirik — APPROVED** [review]
Thanks for the addition! Looks ready to merge once the docstrings are complete and all tests are passing.

**Comment by polavishnu4444** [comment]
Is the CLIPSeg yet to be released in the latest version?
